### PR TITLE
[ML-DataFrame] Fix off-by-one error in checkpoint operations_behind

### PR DESCRIPTION
**Pull request by droberts195**
Fixes a problem where operations_behind would be one less than
expected per shard in a new index matched by the data frame
transform source pattern.

For example, if a data frame transform had a source of foo*
and a new index foo-new was created with 2 shards and 7 documents
indexed in it then operations_behind would be 5 prior to this
change.

The problem was that an empty index has a global checkpoint
number of -1 and the sequence number of the first document that
is indexed into an index is 0, not 1.  This doesn't matter for
indices included in both the last and next checkpoints, as the
off-by-one errors cancelled, but for a new index it affected
the observed result.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/transforms/DataFrameTransformCheckpoint.java [hunk 1]
```diff
@@ -291,8 +291,14 @@ public static long getBehind(DataFrameTransformCheckpoint oldCheckpoint, DataFra
             }
         }
 
-        for (long[] v : newCheckpoint.indicesCheckpoints.values()) {
-            newCheckPointSum += Arrays.stream(v).sum();
+        for (Entry<String, long[]> entry : newCheckpoint.indicesCheckpoints.entrySet()) {
+            newCheckPointSum += Arrays.stream(entry.getValue()).sum();
```
Comment (droberts195) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/transforms/DataFrameTransformCheckpoint.java]: Note that for every brand new shard that has never had a document indexed into it this sum with include a value of -1.  This is not new - the equivalent line in the old code did this too.  Before this PR you could make `operations_behind` go negative by creating a new index with more shards than documents!

**Comment by droberts195** [comment]
Jenkins run elasticsearch-ci/1

**Comment by droberts195** [comment]
Jenkins run elasticsearch-ci/2

**Review by hendrikmuhs — COMMENTED** [review]
Did you considered a +1 for every shard in both sets (old/new checkpoints)? It should have the same effect but seems easier to understand code-wise. As we do not store any of those numbers, BWC is not an issue.

I am unsure about the severity, did you observed a serious issue in 7.3: exception or missed update?
(In 7.3 this is functional, for 7.4 the number is only informational due to the new change-detection.) 
I am fine if you like to backport it, just want to know how severe the issue is.

**Comment by hendrikmuhs** [comment]
run elasticsearch-ci/1

**Comment by droberts195** [comment]
> Did you considered a +1 for every shard in both sets (old/new checkpoints)? It should have the same effect but seems easier to understand code-wise.

I am not sure it would be easier to understand.  The values that are stored for the shards are the global checkpoint numbers for the shards and these range from -1 (no operations on the shard), through 0 (1 operation on the shard), and upwards.  So to create a new number where 0 means no operations on the shard, 1 means operation, etc., would really require a new name, so that we avoid the confusion of having global checkpoint numbers in one part of the code different to another part of the code.  The new name could be `operationsOnShard` or similar, but then that's a whole new concept for somebody who knows what global checkpoint numbers are.

For example, this:

```
        // create the final structure
        Map<String, long[]> checkpointsByIndexReduced = new TreeMap<>();
        checkpointsByIndex.forEach((indexName, checkpoints) -> {
            checkpointsByIndexReduced.put(indexName, checkpoints.values().stream().mapToLong(l -> l).toArray());
        });
```

could change to:

```
        // create the final structure
        Map<String, long[]> operationsByIndex = new TreeMap<>();
        checkpointsByIndex.forEach((indexName, checkpoints) -> {
            operationsByIndex.put(indexName, checkpoints.values().stream().mapToLong(l -> l + 1).toArray());
        });
```

Effectively this is moving from sequence numbers to operations at an earlier stage.  I guess we've already introduced this term "operations" with the `operations_behind` statistic.  So maybe it's OK to propagate it further through our code.  I don't think it's that clear-cut though.  Let's see if anyone else on the team thinks one or the other is clearer for future maintenance.

> did you observed a serious issue in 7.3

When Steve was putting together his demo I think this was affecting him.  Realistically it's more likely to affect demos than production, because (a) it's more likely that there's a desire to trigger an update from a new single document and (b) in demos it's more likely that the document to trigger an update goes in a brand new index so that it's easy to reset the demo environment.

But if we think nobody is ever going to demo continuous transforms on 7.3 again then maybe it's not worth backporting this.

**Comment by hendrikmuhs** [comment]
> > Did you considered a +1 for every shard in both sets (old/new checkpoints)? It should have the same effect but seems easier to understand code-wise.
> 
> I am not sure it would be easier to understand. The values that are stored for the shards are the global checkpoint numbers for the shards and these range from -1 (no operations on the shard), through 0 (1 operation on the shard), and upwards. So to create a new number where 0 means no operations on the shard, 1 means operation, etc., would really require a new name, so that we avoid the confusion of having global checkpoint numbers in one part of the code different to another part of the code. The new name could be `operationsOnShard` or similar, but then that's a whole new concept for somebody who knows what global checkpoint numbers are.

I meant replacing `oldCheckPointSum` and `newCheckPointSum`, only touching code in `getBehind`, not the outer scope. Changing `checkpointsByIndex` would break backwards compatibility.

So something like:

```
// Sequence numbers start at 0 rather than 1, so the number of operations is sequence number + 1
long operationsOldCheckPoints = 0;
long operationsNewCheckPoints = 0;

for (Entry<String, long[]> entry : oldCheckpoint.indicesCheckpoints.entrySet()) {
            // ignore entries that aren't part of newCheckpoint, e.g. deleted indices
            if (newCheckpoint.indicesCheckpoints.containsKey(entry.getKey())) {
                operationsOldCheckPoints += Arrays.stream(entry.getValue()).sum() + entry.getValue().length;
            }
        }
```

**Review by hendrikmuhs — APPROVED** [review]
LGTM

**Comment by droberts195** [comment]
Jenkins run elasticsearch-ci/1

**Pull request merged**